apps/src/lab2/levelEditors/aichatSettings/EditAichatSettings.tsx

```diff
@@ -27,6 +30,50 @@ import ModelSelectionFields from './ModelSelectionFields';
 import VisibilityDropdown from './VisibilityDropdown';
 import {UpdateContext} from './UpdateContext';
 import moduleStyles from './edit-aichat-settings.module.scss';
+import {getTypedKeys} from '@cdo/apps/types/utils';
+
+function sanitizeSettings(settings: LevelAichatSettings) {
+  const sanitizedModelCardInfo = sanitizeField(
+    settings.initialCustomizations.modelCardInfo,
+    EMPTY_MODEL_CARD_INFO
+  );
+  const sanitizedCustomizations = sanitizeField(
+    settings.initialCustomizations,
+    EMPTY_AI_CUSTOMIZATIONS
+  );
+  const sanitizedVisibilities = sanitizeField(
+    settings.visibilities,
+    DEFAULT_VISIBILITIES
+  );
+
+  return {
+    ...settings,
+    initialCustomizations: {
+      ...sanitizedCustomizations,
+      modelCardInfo: sanitizedModelCardInfo,
+    },
+    visibilities: sanitizedVisibilities,
+  };
+}
+
+function sanitizeField<M extends object>(field: M, defaults: M) {
+  const validKeys = getTypedKeys<keyof M>(defaults);
+  // Delete entries for unrecognized keys
+  for (const key of getTypedKeys<keyof M>(field)) {
+    if (!getTypedKeys<keyof M>(defaults).includes(key)) {
+      delete field[key];
+    }
+  }
+
+  // Set default values for missing keys
+  for (const validKey of validKeys) {
+    if (field[validKey] === undefined) {
+      field[validKey] = defaults[validKey];
+    }
+  }
+
+  return field;
+}
```

Comment:
Unless I'm misreading it, this function is impure and will change the original object.

The wrapper `sanitizeSettings` function is passing in a sub-object to the `sanitizeFIeld` function, but _that_ function is destructively iterating over the subobject and deleting keys from it.

At the end of `sanitizeSettings`, it will still return a new wrapper object, but the _original_ `settings` object that was handed in will have its `visibilities` and `initialCustomizations` keys changed in place. Right?

I think just adding:
```
const sanitizedField = {...field}
```

at the start of the function, and changing the later refs from `field` -> `sanitizedField` should fix it.

Also, on line 63, we already have the keys for the default, so can swap this:
```
!getTypedKeys<keyof M>(defaults).includes(key)
```
to this:
```
!validKeys.includes(key)
```

Finally, really just as a note, this is using `Array.includes`, which is _probably_ fine here but is also doing a linear scan of the array every time, and actually hiding that the first loop is O(n^2) and not just O(n).

Since the list of keys is (presumably) very short, it probably doesn't matter. But if the list of keys is "big" and/or this function is called frequently, it might be beneficial to just make a `Set` of `validKeys`:

```
const validKeysSet = new Set(validKeys)
// and then later swap this:
validKeys.includes(key)
// for this:
validKeysSet.has(key)
```

We could grind out a little more performance by yanking out that set creation to an earlier spot, maybe by handing it into the `sanitizeField` function as another arg? Or keeping a memoized copy somewhere else? But I think our returns are starting to diminish at that point.

Reply:
Wait. Ignore everything in that previous comment and do this instead.

That double looping really bugged me, and we can actually get rid of it, solve our overwrite issue, _and_ simplify the function simultaneously. And improve our runtime to O(n) from O(n^2 * n)!

```
function sanitizeField<M extends object>(field: M, defaults: M) {
  // iterate over our default keys. This way, we know that we're _only_ going to keep keys that are valid
  // and we don't need to worry about deleting ones that shouldn't be there
  // we're going to reduce that list into a new object to return
  return getTypedKeys<keyof M>(defaults).reduce(( newField, key) => (
    // and in our reduce, it's easy - since we know that we're only dealing with valid keys, then keep
    // whatever value is in our existing object if present, and otherwise stick in the default.
    { ...newField, [key] : field[key] ?? defaults[key] }
  ), {})
}
```

And I even forewent my usual insistence on naming the accumulator `bucket` for some added clarity. 😄 

I think that'll do it, though of course...
```
"Beware of bugs in the above code; I have only proved it correct, not tried it." - Donald Knuth
```

Reply:
Slick, I love it! Thanks, will update!